### PR TITLE
Tdl 17935 add custom error handling and backoff

### PR DESCRIPTION
**Pull request by namrata270998**
# Description of change

- Added proper error handling based on return codes in the response
- Added retry for the code 50000

# Manual QA steps
 - Check that the tap raises an error with proper error messages in case of error
 - Check that the request is backoff for 50000 error code.
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/unittests/test_error_handling.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        self.headers = headers
+
+    def json(self):
+        return self.text
+
+# function to get mocked response
+def get_response(status_code, json={}, headers=None):
+    return Mockresponse(status_code, json, headers)
+
+# @mock.patch("time.sleep")
```
Comment (savan-chovatiya) [inline tests/unittests/test_error_handling.py:20]: Remove commented lines from the code.

File: tests/unittests/test_error_handling.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+@mock.patch("requests.Session.request")
+class TestInternalErrorBackoff(unittest.TestCase):
```
Comment (savan-chovatiya) [inline tests/unittests/test_error_handling.py:22]: Change the class name as these test cases are for error messages and not for backoff.

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tests/unittests/test_error_handling.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        # set config
+        config = {
+            "access_token": "test_access_token",
+            "user_agent": "test_user_agent"
+        }
+
+        # verify that we raise Timeout error when using "with" statement
+        with self.assertRaises(TikTokAdsClientError) as e:
+            # create client and call function
+            client = TikTokClient(config.get("access_token"), config.get("user_agent"))
```
Comment (umangagrawal-crest) [inline tests/unittests/test_error_handling.py:65]: Move this client initialization out of the `with` statement, so it makes sure that the error is raised from the client.request.
Also, make same changes in other functions wherever it applies

File: tests/unittests/test_error_handling.py (hunk 1 above)
Comment (namrata270998) [reply tests/unittests/test_error_handling.py]: removed the commented line

File: tests/unittests/test_error_handling.py (hunk 3 above)
Comment (namrata270998) [reply tests/unittests/test_error_handling.py]: Updated the code

File: tests/unittests/test_error_handling.py (hunk 2 above)
Comment (namrata270998) [reply tests/unittests/test_error_handling.py]: Changed the class name

File: tap_tiktok_ads/client.py [hunk 4]
```diff
@@ -38,10 +60,21 @@ def check_access_token(self):
         if response.status_code != 200:
             LOGGER.error('Error status_code = %s', response.status_code)
             return False
```
Comment (dbshah1212) [inline tap_tiktok_ads/client.py:62]: Will this raise an error? If not then we should here.

File: tap_tiktok_ads/client.py [hunk 5]
```diff
@@ -1,12 +1,27 @@
...
+    def __init__(self, message=None, response=None):
+        super().__init__(message)
+        self.message = message
+        self.response = response
+
+def should_retry(e):
+    """ Return true if exception is required to retry otherwise return false """
+    response = e.response
+    error_code = response.json().get("code")
+    if error_code == 50000:
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tap_tiktok_ads/client.py:24]: Added comment

**Review by savan-chovatiya — APPROVED** [review]

**Review by dbshah1212 — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

**Pull request merged**